processor.go

```diff
@@ -465,86 +465,91 @@ func (p *ServiceProcessor) ProcessClientStreamRequest(req *http.Request, path st
 		return nil, err
 	}
 
-	var stopServiceChan chan bool
-	var reply interface{}
-
 	// This goroutine listens on any new messages from the client and executes
 	// the request. Executing the request should fill the service's channel, as
 	// the service will use the same chanel for further requests.
+	stopAll := make(chan struct{})
+	closing := sync.Mutex{}
+
 	go func() {
```

Comment:
I'm pretty sure that the channels are not correctly handled here. Some things to take into account:
- every call to `callInterfaceFunc` will create a new `stopServiceChan` - I think the one from the previous call should be closed to indicate that the previous call is done now
- there are some `return`s which will not close the channels

One possibility is to have only one select in the go-routine. But there are some initial conditions that are a bit awkward: `stopServiceChan` and `reply` need to be initialized as empty channels at the beginning, and before closing, the `stopServiceChan` needs always to be tested in a select.

Anyway, I hope it reflects my proposition to close all past calls to `callInterfaceFun` whenever a new input comes in. Also, it reduces the number of go-routines from 3 to 1. And it should correctly handle any errors that are printed and then `return`ed.

Perhaps @tharvik has a better idea how to handle the initial conditions of `stopServiceChan` and `reply`...

```
go func(){
  var stopServiceChan := make chan
  var reply := make chan
  defer func(){
    close(stopServiceChan) // need to use select here
    close(outChan)
  }()
  for {
    select{
    case buf, ok := <- clientInputs:
      if !ok {
        return
      }
      close(stopServiceChan) // that one might work w/o the select
      reply, stopServiceChan, err = callInterfaceFunc...
      if err != nil{
        log.Err()
        return
      }
    case out := <- reply: // I suppose that the reply channel will never be closed. I also suppose you'll find the reflect magic to have an actual channel ;)
      outChan <- reply
    }
  }
}   
```

Reply:
> I think the one from the previous call should be closed to indicate that the previous call is done now

no, because it would prematurely close the `outChan`, which is used among all calls.

>  there are some returns which will not close the channels

please elaborate

Reply:
> no, because it would prematurely close the `outChan`, which is used among all calls.

In your code this is true. But in the select I propose above, this wouldn't happen anymore.

If you don't close the previous `stopServiceChan`, every time a user sends a message through an existing connection you'll have one more `callInterfaceFun` sending data. Which is probably not a problem for the `paginate` service, as it stops after final block is sent anyway. But for a service like `StreamingTransaction`, if the user sends a message to an existing connection, you'll have two go-routines sending block updates to the client.

So it looks cleaner and simpler to me to have only one `callInterfaceFun` active at any one time by closing the `stopServiceChan`.

> please elaborate

Any of the `return`s in the beginning of the `for`-loop will just silently quit and not close the `outChan` to indicate that the connection is done.